### PR TITLE
Safer safelok routines

### PR DESCRIPTION
**Pull request by henrygab**
Generally, some bug fixes, and some syntactic sugar (with minimal or no impact to compiled binary).

* FIX: encoding of interval (aka expiration date) was incorrect
* FIX: `const` correctness allows more compiler optimizations.
* Rather than `uint8_t *` of unknown length, functions now take pointers to structures of the appropriate size.
  * Makes it easier to properly use the functions, as they self-document not only the type, but how many bytes might be accessed through that pointer.
  * Especially useful when code is later maintained by someone other than original author.
  * May improve some static analysis tools outputs.
* Wrappers for property getters / setters
  * Allows easier single-point-of-correction.
  * Example: When eltrick determined that the card creation year field is only 7-bits, the [fix](https://github.com/RfidResearchGroup/proxmark3/commit/02056c95155214506b8bc4c1153069d4186f245f) was localized.
* **_Very_** basic self-test option added.
  * Framework makes it trivial to add more test cases later.
* Bit set / get routines now reject invalid values
  * Values must fit within number of bits
  * Warning/error messages are output
  * Yes, this helped catch code errors
* Adding missing output for cards that do not have variable keys.
  * No, I don't know what "variable keys" means ... I just fixed the output.

### CODE REVIEW

File: client/src/pm3line_vocabulary.h [hunk 1]
```diff
@@ -359,7 +359,7 @@ const static vocabulary_t vocabulary[] = {
     { 0, "hf mf fchk" },
     { 1, "hf mf decrypt" },
     { 0, "hf mf supercard" },
-    { 1, "hf mf bambukeys" },
+    { 1, "hf mf keygen" },
```
[the remark this answers is not in the document]
Comment (henrygab) [reply client/src/pm3line_vocabulary.h:362]: This was automated change.   Not part of this PR.

File: client/src/cmdhfsaflok.c [hunk 2]
```diff
@@ -30,9 +33,567 @@
...
+    static saflok_mfc_datetime_t add_offset(
+        const saflok_mfc_datetime_t *base,
+        const saflok_mfc_datetime_offset_t offset
+        ) {
+        // adding dates is non-trivial.
+        // luckily, there's already a standard for this...
+        struct tm tm = {
+            .tm_year = base->year - 1900,
+            .tm_mon  = base->month - 1,
+            .tm_mday = base->day - 1,
```
[the remark this answers is not in the document]
Comment (henrygab) [reply client/src/cmdhfsaflok.c]: Fixed in https://github.com/RfidResearchGroup/proxmark3/pull/3026/commits/c98bf9ad3ad5a165e346dd074015c71c531b6f6f

File: client/src/cmdhfsaflok.c [hunk 3]
```diff
@@ -369,66 +901,75 @@ static void saflok_decode(uint8_t *data) {
...
+        // return value ignored ... function here is void(!)
+        set_saflok_mfc_card_creation_datetime(data, &creation_dt);
+    }
+
+    saflok_mfc_datetime_t expiration_dt = {0};
+    if (sscanf(dt_expiration, fmt, &expiration_dt.year, &expiration_dt.month, &expiration_dt.day, &expiration_dt.hour, &expiration_dt.minute) == 5) {
+        // return value ignored ... function here is void(!)
+        set_saflok_mfc_card_expiration_datetime(data, &expiration_dt);
+    }
+
```
[the remark this answers is not in the document]
Comment (henrygab) [reply client/src/cmdhfsaflok.c]: Fixed in https://github.com/RfidResearchGroup/proxmark3/pull/3026/commits/c98bf9ad3ad5a165e346dd074015c71c531b6f6f

File: client/src/cmdhfsaflok.c [hunk 4]
```diff
@@ -30,9 +33,567 @@
...
+            PrintAndLogEx(WARNING, "Warning: interval days appears invalid (%u)\n", days);
+        }
+        return days;
+    }
+    static inline bool _set_saflok_mfc_interval_days_impl(saflok_mfc_data_t *data, uint8_t days) {
+        if (days > 31u) {
+            PrintAndLogEx(ERR, _RED_("ERROR:") " interval days out of range (%u)\n", days);
+            return false;
+        }
+        data->raw[9] = (data->raw[9] & 0xC7u) | ((days & 0x1Fu) << 3);
```
[the remark this answers is not in the document]
Comment (henrygab) [reply client/src/cmdhfsaflok.c]: Fixed in https://github.com/RfidResearchGroup/proxmark3/pull/3026/commits/c98bf9ad3ad5a165e346dd074015c71c531b6f6f

File: client/src/cmdhfsaflok.c [hunk 5]
```diff
@@ -106,259 +669,228 @@ uint8_t magic_table[192] = {
...
+static void set_bitfield(saflok_mfc_data_t *data, size_t start_bit, size_t num_bits, uint32_t value) {
+    static const size_t total_available_bits = ARRAYLEN(data->raw) * 8;
+    if (start_bit >= total_available_bits) {
+        // Out of bounds access
+        PrintAndLogEx(ERR, "set_bitfield: out of bounds access (start_bit=%zu, total_available_bits=%zu)\n", start_bit, total_available_bits);
+        return;
+    }
+    if (num_bits > 32) {
+        // Exceeds maximum supported bit extraction
+        PrintAndLogEx(ERR, "set_bitfield: num_bits exceeds 32\n", num_bits);
```
[the remark this answers is not in the document]
Comment (henrygab) [reply client/src/cmdhfsaflok.c]: Fixed in https://github.com/RfidResearchGroup/proxmark3/pull/3026/commits/c98bf9ad3ad5a165e346dd074015c71c531b6f6f

File: client/src/cmdhfsaflok.c [hunk 6]
```diff
@@ -30,9 +33,567 @@
...
+            return false;
+        }
+        data->raw[11] = (data->raw[11] & 0xF0u) | (month & 0x0Fu);
+        return true;
+    }
+    static inline uint8_t _get_saflok_mfc_card_creation_day_impl(const saflok_mfc_data_t *data) {
+        return ((data->raw[12] >> 3) & 0x1Fu);
+    }
+    static inline bool _set_saflok_mfc_card_creation_day_impl(saflok_mfc_data_t *data, uint8_t day) {
+        data->raw[12] = (data->raw[12] & 0xE0u) | ((day & 0x1Fu) << 3);
```
[the remark this answers is not in the document]
Comment (henrygab) [reply client/src/cmdhfsaflok.c]: Fixed in https://github.com/RfidResearchGroup/proxmark3/pull/3026/commits/c98bf9ad3ad5a165e346dd074015c71c531b6f6f

File: client/src/cmdhfsaflok.c [hunk 7]
```diff
@@ -106,259 +669,228 @@ uint8_t magic_table[192] = {
...
+
     for (size_t i = 0; i < num_bits; i++) {
         size_t current_bit = start_bit + i;
         size_t byte_index = current_bit / 8;
         size_t bit_index = 7 - (current_bit % 8);
 
         uint32_t bit_value = (value >> (num_bits - 1 - i)) & 1U;
 
-        data[byte_index] = (data[byte_index] & ~(1 << bit_index)) | (bit_value << bit_index);
+        data->raw[byte_index] = (data->raw[byte_index] & ~(1 << bit_index)) | (bit_value << bit_index);
```
[the remark this answers is not in the document]
Comment (henrygab) [reply client/src/cmdhfsaflok.c:981]: I need to verify if GitHub Copilot's comments are accurate or not.

File: client/src/cmdhfsaflok.c [hunk 8]
```diff
@@ -30,9 +33,567 @@
...
+            return false;
+        }
+        set_bitfield(data, 88, 28, raw_card_creation_date);
+        return true;
+    }
+    static inline uint16_t get_saflok_mfc_property_id(const saflok_mfc_data_t *data) {
+        return get_bitfield(data, 116, 12);
+    }
+    static inline bool set_saflok_mfc_property_id(saflok_mfc_data_t *data, uint32_t property_id) {
+        if (property_id > 0xFFFFu) {
```
[the remark this answers is not in the document]
Comment (henrygab) [reply client/src/cmdhfsaflok.c]: Fixed in https://github.com/RfidResearchGroup/proxmark3/pull/3026/commits/c98bf9ad3ad5a165e346dd074015c71c531b6f6f

File: client/src/cmdhfsaflok.c [hunk 9]
```diff
@@ -30,9 +33,567 @@
...
+        }
+        if (tm.tm_mon >= 12) {
+            tm.tm_year ++;
+            tm.tm_mon -= 12;
+        }
+        // and finally convert back into saflok_mfc_datetime_t
+        saflok_mfc_datetime_t result = {
+            .year   = tm.tm_year + 1900,
+            .month  = tm.tm_mon + 1,
+            .day    = tm.tm_mday + 1,
```
[the remark this answers is not in the document]
Comment (henrygab) [reply client/src/cmdhfsaflok.c]: Fixed in https://github.com/RfidResearchGroup/proxmark3/pull/3026/commits/c98bf9ad3ad5a165e346dd074015c71c531b6f6f

File: doc/commands.json [hunk 10]
```diff
@@ -3256,7 +3256,7 @@
...
.... } iclass { ICLASS RFIDs... } ict { ICT MFC/DESfire RFIDs... } jooki { Jooki RFIDs... } ksx6924 { KS X 6924 (T-Money, Snapper+) RFIDs } legic { LEGIC RFIDs... } lto { LTO Cartridge Memory RFIDs... } mf { MIFARE RFIDs... } mfp { MIFARE Plus RFIDs... } mfu { MIFARE Ultralight RFIDs... } mfdes { MIFARE Desfire RFIDs... } ntag424 { NXP NTAG 4242 DNA RFIDs... } saflok { Saflok MFC RFIDs... } seos { SEOS RFIDs... } st25ta { ST25TA RFIDs... } tesla { TESLA Cards... } texkom { Texkom RFIDs... } thinfilm { Thinfilm RFIDs... } topaz { TOPAZ (NFC Type 1) RFIDs... } vas { Apple Value Added Service... } xerox { Fuji/Xerox cartridge RFIDs... } ----------- --------------------- General --------------------- help This help list List protocol data in trace buffer search Search for known HF tags --------------------------------------------------------------------------------------- hf list available offline: yes Alias of `trace list -t raw` with selected protocol data to annotate trace buffer You can load a trace from file (see `trace load -h`) or it be downloaded from device by default It accepts all other arguments of `trace list`. Note that some might not be relevant for this specific protocol",
```
[the remark this answers is not in the document]
Comment (henrygab) [reply doc/commands.json:3259]: not part of this PR ... automated `make style` change

**Comment by iceman1001** [comment]
Interesting,    just after a quick look I can see some issue.
1.  make style,   your setup doesn't have waveshare meaning we loose some commands now.   Please revert.
2. uid can be 4,7,10   there is nothing saying a MFC for saflok can't be a UID 7 bytes.  

Some people love structs,   as long as they fit all the scenarios it's ok.   We tend to use uid[10]  and uidlen in the struct to keep track of things. 

3.  assert,  as long as  assert doesn't "exit the client" when triggered its ok,  but if not the client isn't handled in a nice way.   Then you have to redo the concept.

File: client/src/cmdhfsaflok.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (henrygab) [reply client/src/cmdhfsaflok.c:981]: After adding [explicit self-test](https://github.com/RfidResearchGroup/proxmark3/pull/3026/commits/4cb20bde5fbc4c0b41e6800537f122a98161a485) of `get_bitfield()` and `set_bitfield()`, I'm calling this one a false-positive.  
That self-test exercises a broad ranges of values and does not find this inconsistency.

I'll leave this unresolved for now ... would like a second set of eyes.

**Comment by henrygab** [comment]
> 1. make style,   your setup doesn't have waveshare meaning we loose some commands now.   Please revert.

Interesting... I cannot both test my changes and easily revert this, but as it's automated, perhaps you could simply run the automated command after the merge?  I have no idea why the waveshare command isn't available for my builds.   If you know why, I'd appreciate if you could share the secret!

> 2. uid can be 4,7,10   there is nothing saying a MFC for saflok can't be a UID 7 bytes.

While the MFC UID might be longer, the algorithm as used in the code I edited was limiting the UID to 4 bytes.  This code replicated the original behavior, and I do not have a MFC Saflok with 7 or 10 byte UID to test/validate such an algorithmic change.

> Some people love structs, as long as they fit all the scenarios it's ok. We tend to use uid[10] and uidlen in the struct to keep track of things.

The saflok code that was checked in used only the first four bytes of the UID (even though it read more bytes from the card).  I did not change the logic   When this command is expanded to support >4 bytes of the UID (or UL-C, or any of a number of other extensions), adding an explicit length field for the UID **_makes total sense and absolutely should be done_**.

> 3. assert,  as long as  assert doesn't "exit the client" when triggered its ok,  but if not the client isn't handled in a nice way.   Then you have to redo the concept.

There ~~are~~ **were** only two `assert()` in the code.   ~~Yes, this exits the client.  These two asserts are for programming errors (not user / data errors).  There is no other option in the current code's architecture to prevent either silent data loss, silent data corruption, or silently returning bad data to the caller ... which is likely to be data corruption or data loss.~~

<hr/>
<details><summary>Are you sure you don't want *_any_* asserts?</summary><p/>

I agree that user errors should not be assertions.
I also agree that receiving corrupt or unexpected data from a card should not cause an assertion.

At the same time, asserts catch **_coder/programmer_** errors.  Unfortunately, much of the code does not have proper error handling ... or explicitly ignores error return values.   Thus, there is often no way to return an error cleanly from a deep stack, because intermediate functions will blindly ignore the error and continue, even when something went wrong.

Prior to adding the validation, some of the behavior was (iirc) silent data corruption.  If lucky, this would crash the client.  If unlucky, it could corrupt the data on the card.   The silent nature of that data corruption is not something I think should be left to occur.

Therefore, what behavior would you suggest for the following situation:
* the error is a code bug (e.g., setting values that don't fit in the bitfield length)
* the error in the original code base would not be handled/propagated correctly
* the error in the original code base might silently corrupt data

In the end, you have final choice to accept or reject the PR, and can make changes to the PR (before or after merging).  In other words... you have absolute choice...

</details><hr/>

**Comment by iceman1001** [comment]
Take my opinions with the perspective of having a nice user experience.   The client randomly exiting is not a nice user experience.  
When an error is detected,   clean up as much as we can after it and let the user enter new commands in the CLI.    As robust as we can.

**Comment by henrygab** [comment]
I'm open to you changing the asserts in the saflok file.   Of course, there are other `assert()` in the client code that should also be changed, if this is the goal:
* `client/src/pm3_luawrap.c`
* `client/src/imgutils.c`
* `client/src/py3_pywrap.c`
* `client/src/fido/cbortools.c`
* `common/lz4/lz4.c`
* ...

I'm off to bed.  I've fixed enough issues (esp. around interval / expiration) that I recommend merging, and then changing whatever remains that you don't like.

**Comment by henrygab** [comment]
And ... I've added some regressions in date handling.

**Comment by henrygab** [comment]
@iceman1001 -- Here's the link to the KDF algorithm prior to my changes.  Note it only uses / supports a 4-byte UID.

https://github.com/STIEBELJOSHUA/proxmark3/blob/15026f7b3506b75b47831d75fa66bc25efc5ca2e/client/src/cmdhfsaflok.c#L282-L297

I simply wanted you to see that I wasn't trying to remove support for other UID sizes ... the algorithm simply didn't use / support anything else, and I don't have the knowledge on changes that would be needed for other UID sizes (or any ability to test them).

**Comment by henrygab** [comment]
@iceman1001 -- asserts removed, manually tested both valid and invalid edge cases for expiration dates. There's still differences between `hf mf info` and `hf saflok read` for certain expiration dates, but I believe the bug lies in `hf mf info`.
Pending completed CI checks, this is ready.
